File: cpp/src/provider/mobygames.cpp
/// @file mobygames.cpp
/// @brief MobyGames metadata provider implementation

#include <retro_metadata/cache/cache.hpp>
#include <retro_metadata/config.hpp>
#include <retro_metadata/errors.hpp>
#include <retro_metadata/internal/matching.hpp>
#include <retro_metadata/internal/normalization.hpp>
#include <retro_metadata/provider/provider.hpp>
#include <retro_metadata/provider/registry.hpp>

#include <cpr/cpr.h>
#include <nlohmann/json.hpp>

#include <regex>

namespace retro_metadata {

namespace {

/// @brief Regex to match MobyGames ID tags in filenames like (moby-12345)
const std::regex kMobyGamesTagRegex(R"(\(moby-(\d+)\))", std::regex::icase);

/// @brief Regex to match Sony serial codes like SLUS-12345, SCUS-97328
const std::regex kSonySerialRegex(R"(([A-Z]{4})[_-](\d{5}))", std::regex::icase);

/// @brief Regex to match PS2 OPL format like SLUS_123.45
const std::regex kPS2OPLRegex(R"(([A-Z]{4})_(\d{3})\.(\d{2}))", std::regex::icase);

/// @brief Regex to match Nintendo Switch 16-character hex IDs
const std::regex kSwitchTitleDBRegex(R"(([0-9A-Fa-f]{16}))");

/// @brief Regex to match Switch product IDs like LA-H-AAAAA
const std::regex kSwitchProductIDRegex(R"([A-Z]{2}-[A-Z]-([A-Z0-9]{5}))", std::regex::icase);

/// @brief Regex to match MAME ROM names
const std::regex kMAMEArcadeRegex(R"(^[a-z0-9_]+$)", std::regex::icase);

/// @brief Base URL for the MobyGames API
constexpr const char* kBaseURL = "https://api.mobygames.com/v1";

/// @brief Safely gets a string from a JSON object
std::string get_string(const nlohmann::json& j, const std::string& key) {
    if (j.contains(key) && j[key].is_string()) {
        return j[key].get<std::string>();
    }
    return "";
}

/// @brief Safely gets a number from a JSON object
double get_number(const nlohmann::json& j, const std::string& key) {
    if (j.contains(key) && j[key].is_number()) {
        return j[key].get<double>();
    }
    return 0.0;
}

/// @brief Cleans a filename by removing extension and tags
std::string clean_filename(const std::string& filename) {
    static const std::regex ext_pattern(R"(\.[^.]+$)");
    static const std::regex tag_pattern(R"(\s*[\(\[][^\)\]]*[\)\]])");
    std::string name = std::regex_replace(filename, ext_pattern, "");
    name = std::regex_replace(name, tag_pattern, "");
    auto start = name.find_first_not_of(" \t\n\r");
    if (start == std::string::npos) return "";
    auto end = name.find_last_not_of(" \t\n\r");
    return name.substr(start, end - start + 1);
}

/// @brief Extracts Sony serial code from filename (PS1/PS2/PSP)
std::string extract_serial_code(const std::string& filename) {
    std::smatch match;

    // Try PS2 OPL format first (SLUS_123.45)
    if (std::regex_search(filename, match, kPS2OPLRegex) && match.size() > 3) {
        std::string prefix = match[1].str();
        std::transform(prefix.begin(), prefix.end(), prefix.begin(), ::toupper);
        return prefix + "-" + match[2].str() + match[3].str();
    }

    // Try standard Sony serial format (SLUS-12345 or SLUS_12345)
    if (std::regex_search(filename, match, kSonySerialRegex) && match.size() > 2) {
        std::string prefix = match[1].str();
        std::transform(prefix.begin(), prefix.end(), prefix.begin(), ::toupper);
        return prefix + "-" + match[2].str();
    }

    return "";
}

/// @brief Extracts Nintendo Switch product ID from filename
std::string extract_switch_product_id(const std::string& filename) {
    std::smatch match;
    if (std::regex_search(filename, match, kSwitchProductIDRegex) && match.size() > 1) {
        std::string product_id = match[1].str();
        std::transform(product_id.begin(), product_id.end(), product_id.begin(), ::toupper);
        return product_id;
    }
    return "";
}

/// @brief Checks if filename is in MAME format
bool is_mame_format(const std::string& filename) {
    static const std::regex ext_pattern(R"(\.[^.]+$)");
    std::string name = std::regex_replace(filename, ext_pattern, "");
    // MAME names are typically short (under 20 chars) and alphanumeric
    if (name.length() > 20) {
        return false;
    }
    return std::regex_match(name, kMAMEArcadeRegex);
}

/// @brief MobyGames platform IDs for special handling
constexpr int kPlatformPS1 = 6;
constexpr int kPlatformPS2 = 7;
constexpr int kPlatformPSP = 46;
constexpr int kPlatformArcade = 143;
constexpr int kPlatformSwitch = 203;

}  // namespace

/// @brief MobyGames metadata provider
class MobyGamesProvider : public Provider {
public:
    MobyGamesProvider(const ProviderConfig& config, std::shared_ptr<Cache> cache)
        : config_(config), cache_(std::move(cache)) {}

    std::string name() const override { return "mobygames"; }

    std::vector<SearchResult> search(const std::string& query, const SearchOptions& opts) override {
        if (!config_.is_configured()) return {};

        cpr::Parameters params{
            {"title", query},
            {"api_key", config_.get_credential("api_key")}
        };

        int limit = opts.limit > 0 ? opts.limit : 10;
        params.Add({"limit", std::to_string(limit)});

        if (opts.platform_id) {
            params.Add({"platform", std::to_string(*opts.platform_id)});
        }

        auto response = make_request("/games", params);
        if (!response.contains("games") || !response["games"].is_array()) {
            return {};
        }

        std::vector<SearchResult> results;
        for (const auto& game : response["games"]) {
            SearchResult sr;
            sr.provider = name();
            sr.provider_id = static_cast<int>(get_number(game, "game_id"));
            sr.name = get_string(game, "title");

            // Extract cover URL from sample_cover
            if (game.contains("sample_cover") && game["sample_cover"].is_object()) {
                sr.cover_url = get_string(game["sample_cover"], "image");
            }

            // Extract platforms
            if (game.contains("platforms") && game["platforms"].is_array()) {
                for (const auto& pl : game["platforms"]) {
                    sr.platforms.push_back(get_string(pl, "platform_name"));
                }
            }

            // Extract release year from first platform
            if (game.contains("platforms") && game["platforms"].is_array() &&
                !game["platforms"].empty()) {
                std::string date_str = get_string(game["platforms"][0], "first_release_date");
                if (date_str.length() >= 4) {
                    try {
                        sr.release_year = std::stoi(date_str.substr(0, 4));
                    } catch (...) {
                        // Ignore parse errors
                    }
                }
            }

            results.push_back(std::move(sr));
        }

        return results;
    }

    std::unique_ptr<GameResult> get_by_id(int game_id) override {
        if (!config_.is_configured()) return nullptr;

        cpr::Parameters params{
            {"api_key", config_.get_credential("api_key")}
        };

        auto response = make_request("/games/" + std::to_string(game_id), params);
        if (get_number(response, "game_id") == 0) {
            return nullptr;
        }

        return build_game_result(response);
    }

    std::unique_ptr<GameResult> identify(
        const std::string& filename, const IdentifyOptions& opts) override {
        if (!config_.is_configured()) return nullptr;

        // Check for MobyGames ID tag in filename
        std::smatch match;
        if (std::regex_search(filename, match, kMobyGamesTagRegex) && match.size() > 1) {
            try {
                int id = std::stoi(match[1].str());
                auto result = get_by_id(id);
                if (result) return result;
            } catch (...) {
                // Ignore parse errors, continue with filename matching
            }
        }

        if (!opts.platform_id) return nullptr;

        int platform_id = *opts.platform_id;
        std::string search_term;

        // Try Sony serial format for PS1/PS2/PSP platforms
        if (platform_id == kPlatformPS1 || platform_id == kPlatformPS2 ||
            platform_id == kPlatformPSP) {
            std::string serial = extract_serial_code(filename);
            if (!serial.empty()) {
                search_term = serial;
            }
        }

        // Try Nintendo Switch product ID format
        if (platform_id == kPlatformSwitch && search_term.empty()) {
            std::string product_id = extract_switch_product_id(filename);
            if (!product_id.empty()) {
                search_term = product_id;
            }
        }

        // Try MAME format for arcade platform
        if (platform_id == kPlatformArcade && search_term.empty()) {
            if (is_mame_format(filename)) {
                static const std::regex ext_pattern(R"(\.[^.]+$)");
                search_term = std::regex_replace(filename, ext_pattern, "");
            }
        }

        // Fall back to cleaned filename
        if (search_term.empty()) {
            search_term = clean_filename(filename);
        }

        // Search for the game
        cpr::Parameters params{
            {"title", search_term},
            {"platform", std::to_string(platform_id)},
            {"api_key", config_.get_credential("api_key")}
        };

        auto response = make_request("/games", params);

        nlohmann::json games;
        if (response.contains("games") && response["games"].is_array()) {
            games = response["games"];
        }

        // If no results, try splitting by special characters
        if (games.empty() || !games.is_array()) {
            auto terms = normalization::split_search_term(search_term);
            if (terms.size() > 1) {
                params = cpr::Parameters{
                    {"title", terms.back()},
                    {"platform", std::to_string(platform_id)},
                    {"api_key", config_.get_credential("api_key")}
                };
                response = make_request("/games", params);
                if (response.contains("games") && response["games"].is_array()) {
                    games = response["games"];
                }
            }
        }

        if (games.empty() || !games.is_array()) {
            return nullptr;
        }

        // Build map of games by name for matching
        std::map<std::string, nlohmann::json> games_by_name;
        std::vector<std::string> names;
        for (const auto& g : games) {
            std::string title = get_string(g, "title");
            if (!title.empty()) {
                games_by_name[title] = g;
                names.push_back(title);
            }
        }

        // Find best match using Jaro-Winkler similarity
        matching::FindBestMatchOptions match_opts;
        match_opts.min_similarity_score = 0.6;  // MobyGames uses lower threshold
        auto best = matching::find_best_match(search_term, names, match_opts);

        if (!best.found()) {
            return nullptr;
        }

        auto result = build_game_result(games_by_name[best.match]);
        result->match_score = best.score;
        return result;
    }

    void heartbeat() override {
        cpr::Parameters params{
            {"limit", "1"},
            {"api_key", config_.get_credential("api_key")}
        };

        // This will throw appropriate errors if the API is not accessible
        make_request("/games", params);
    }

    void close() override {
        // No resources to clean up
    }

private:
    /// @brief Makes an HTTP request to the MobyGames API
    nlohmann::json make_request(const std::string& endpoint, const cpr::Parameters& params) {
        std::string url = std::string(kBaseURL) + endpoint;

        cpr::Response r = cpr::Get(
            cpr::Url{url},
            params,
            cpr::Header{
                {"Accept", "application/json"},
                {"User-Agent", "retro-metadata/1.0"}
            },
            cpr::Timeout{config_.timeout * 1000}
        );

        if (r.status_code == 401) {
            throw AuthError(name());
        }

        if (r.status_code == 429) {
            throw RateLimitError(name());
        }

        if (r.status_code != 200) {
            throw ConnectionError(name(), "HTTP " + std::to_string(r.status_code));
        }

        try {
            return nlohmann::json::parse(r.text);
        } catch (const nlohmann::json::parse_error& e) {
            throw ConnectionError(name(), "Failed to parse JSON response");
        }
    }

    /// @brief Builds a GameResult from a MobyGames API response
    std::unique_ptr<GameResult> build_game_result(const nlohmann::json& game) {
        auto result = std::make_unique<GameResult>();
        int game_id = static_cast<int>(get_number(game, "game_id"));

        result->provider = name();
        result->provider_id = game_id;
        result->provider_ids = {{"mobygames", game_id}};
        result->name = get_string(game, "title");
        result->summary = get_string(game, "description");

        // Extract cover URL from sample_cover
        if (game.contains("sample_cover") && game["sample_cover"].is_object()) {
            result->artwork.cover_url = get_string(game["sample_cover"], "image");
        }

        // Extract screenshots from sample_screenshots
        if (game.contains("sample_screenshots") && game["sample_screenshots"].is_array()) {
            for (const auto& s : game["sample_screenshots"]) {
                std::string img_url = get_string(s, "image");
                if (!img_url.empty()) {
                    result->artwork.screenshot_urls.push_back(img_url);
                }
            }
        }

        // Extract metadata
        result->metadata = extract_metadata(game);
        result->raw_response = game;

        return result;
    }

    /// @brief Extracts metadata from a MobyGames game object
    GameMetadata extract_metadata(const nlohmann::json& game) {
        GameMetadata metadata;
        metadata.raw_data = game;

        // Genres
        if (game.contains("genres") && game["genres"].is_array()) {
            for (const auto& g : game["genres"]) {
                std::string genre_name = get_string(g, "genre_name");
                if (!genre_name.empty()) {
                    metadata.genres.push_back(genre_name);
                }
            }
        }

        // Alternative names
        if (game.contains("alternate_titles") && game["alternate_titles"].is_array()) {
            for (const auto& t : game["alternate_titles"]) {
                std::string title = get_string(t, "title");
                if (!title.empty()) {
                    metadata.alternative_names.push_back(title);
                }
            }
        }

        // Platforms
        if (game.contains("platforms") && game["platforms"].is_array()) {
            for (const auto& pl : game["platforms"]) {
                Platform platform;
                platform.name = get_string(pl, "platform_name");
                int platform_id = static_cast<int>(get_number(pl, "platform_id"));
                platform.provider_ids = {{"mobygames", platform_id}};
                metadata.platforms.push_back(platform);

                // Extract release year from first platform if not set
                if (!metadata.release_year) {
                    std::string date_str = get_string(pl, "first_release_date");
                    if (date_str.length() >= 4) {
                        try {
                            metadata.release_year = std::stoi(date_str.substr(0, 4));
                        } catch (...) {
                            // Ignore parse errors
                        }
                    }
                }
            }
        }

        // Rating (MobyGames scores are out of 10, convert to 100)
        double moby_score = get_number(game, "moby_score");
        if (moby_score > 0) {
            metadata.total_rating = moby_score * 10;
        }

        return metadata;
    }

    ProviderConfig config_;
    std::shared_ptr<Cache> cache_;
};

namespace {
[[maybe_unused]] ProviderRegistrar mobygames_registrar(
    "mobygames",
    [](const ProviderConfig& config, std::shared_ptr<Cache> cache) {
        return std::make_unique<MobyGamesProvider>(config, std::move(cache));
    });
}

}  // namespace retro_metadata

File: pkg/provider/mobygames/mobygames.go
// Package mobygames provides the MobyGames metadata provider implementation.
package mobygames

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/josegonzalez/retro-metadata/pkg/cache"
	"github.com/josegonzalez/retro-metadata/pkg/internal/normalization"
	"github.com/josegonzalez/retro-metadata/pkg/platform"
	"github.com/josegonzalez/retro-metadata/pkg/provider"
	"github.com/josegonzalez/retro-metadata/pkg/retrometadata"
)

// MobyGamesTagRegex matches MobyGames ID tags in filenames like (moby-12345)
var MobyGamesTagRegex = regexp.MustCompile(`(?i)\(moby-(\d+)\)`)

// SonySerialRegex matches PS1/PS2/PSP serial codes like SLUS-12345, SCUS-97328
var SonySerialRegex = regexp.MustCompile(`(?i)([A-Z]{4})[_-](\d{5})`)

// PS2OPLRegex matches PS2 OPL format like SLUS_123.45
var PS2OPLRegex = regexp.MustCompile(`(?i)([A-Z]{4})_(\d{3})\.(\d{2})`)

// SwitchTitleDBRegex matches Nintendo Switch 16-character hex IDs
var SwitchTitleDBRegex = regexp.MustCompile(`([0-9A-Fa-f]{16})`)

// SwitchProductIDRegex matches Switch product IDs like LA-H-AAAAA
var SwitchProductIDRegex = regexp.MustCompile(`(?i)[A-Z]{2}-[A-Z]-([A-Z0-9]{5})`)

// MAMEArcadeRegex matches MAME ROM names
var MAMEArcadeRegex = regexp.MustCompile(`(?i)^([a-z0-9_]+)$`)

// Provider implements the MobyGames metadata provider.
type Provider struct {
	*provider.BaseProvider
	baseURL    string
	userAgent  string
	httpClient *http.Client
}

// NewProvider creates a new MobyGames provider instance.
func NewProvider(config retrometadata.ProviderConfig, c cache.Cache) (*Provider, error) {
	p := &Provider{
		BaseProvider: provider.NewBaseProvider("mobygames", config, c),
		baseURL:      "https://api.mobygames.com/v1",
		userAgent:    "retro-metadata/1.0",
		httpClient:   &http.Client{Timeout: 30 * time.Second},
	}
	p.SetMinSimilarityScore(0.6)
	return p, nil
}

func (p *Provider) apiKey() string {
	return p.GetCredential("api_key")
}

func (p *Provider) request(ctx context.Context, endpoint string, params map[string]string) (interface{}, error) {
	u, err := url.Parse(p.baseURL + endpoint)
	if err != nil {
		return nil, fmt.Errorf("failed to parse URL: %w", err)
	}

	q := u.Query()
	q.Set("api_key", p.apiKey())
	for k, v := range params {
		q.Set(k, v)
	}
	u.RawQuery = q.Encode()

	req, err := http.NewRequestWithContext(ctx, "GET", u.String(), nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %w", err)
	}
	req.Header.Set("User-Agent", p.userAgent)

	resp, err := p.httpClient.Do(req)
	if err != nil {
		return nil, &retrometadata.ProviderError{Provider: p.Name(), Err: retrometadata.ErrProviderConnection}
	}
	defer resp.Body.Close()

	if resp.StatusCode == 401 {
		return nil, &retrometadata.ProviderError{Provider: p.Name(), Err: retrometadata.ErrProviderAuth}
	}

	if resp.StatusCode == 429 {
		return nil, &retrometadata.ProviderError{Provider: p.Name(), Err: retrometadata.ErrProviderRateLimit}
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response: %w", err)
	}

	var result interface{}
	if err := json.Unmarshal(body, &result); err != nil {
		return nil, fmt.Errorf("failed to parse response: %w", err)
	}

	return result, nil
}

// Search searches for games by name.
func (p *Provider) Search(ctx context.Context, query string, opts retrometadata.SearchOptions) ([]retrometadata.SearchResult, error) {
	if !p.IsEnabled() {
		return nil, nil
	}

	params := map[string]string{
		"title": query,
		"limit": strconv.Itoa(max(opts.Limit, 10)),
	}

	if opts.PlatformID != nil {
		params["platform"] = strconv.Itoa(*opts.PlatformID)
	}

	result, err := p.request(ctx, "/games", params)
	if err != nil {
		return nil, err
	}

	resultMap, ok := result.(map[string]interface{})
	if !ok {
		return nil, nil
	}

	games, ok := resultMap["games"].([]interface{})
	if !ok {
		return nil, nil
	}

	var searchResults []retrometadata.SearchResult
	for _, g := range games {
		game, ok := g.(map[string]interface{})
		if !ok {
			continue
		}

		sr := retrometadata.SearchResult{
			Provider:   p.Name(),
			ProviderID: int(getFloat64(game, "game_id")),
			Name:       getString(game, "title"),
		}

		// Extract cover URL
		if sampleCover, ok := game["sample_cover"].(map[string]interface{}); ok {
			sr.CoverURL = getString(sampleCover, "image")
		}

		// Extract platforms
		if platforms, ok := game["platforms"].([]interface{}); ok {
			for _, pl := range platforms {
				if plMap, ok := pl.(map[string]interface{}); ok {
					sr.Platforms = append(sr.Platforms, getString(plMap, "platform_name"))
				}
			}
		}

		// Extract release year
		if platforms, ok := game["platforms"].([]interface{}); ok && len(platforms) > 0 {
			if firstPlatform, ok := platforms[0].(map[string]interface{}); ok {
				dateStr := getString(firstPlatform, "first_release_date")
				if len(dateStr) >= 4 {
					if year, err := strconv.Atoi(dateStr[:4]); err == nil {
						sr.ReleaseYear = &year
					}
				}
			}
		}

		searchResults = append(searchResults, sr)
	}

	return searchResults, nil
}

// GetByID gets game details by MobyGames ID.
func (p *Provider) GetByID(ctx context.Context, gameID int) (*retrometadata.GameResult, error) {
	if !p.IsEnabled() {
		return nil, nil
	}

	result, err := p.request(ctx, fmt.Sprintf("/games/%d", gameID), nil)
	if err != nil {
		return nil, err
	}

	game, ok := result.(map[string]interface{})
	if !ok || getFloat64(game, "game_id") == 0 {
		return nil, nil
	}

	return p.buildGameResult(game), nil
}

// Identify identifies a game from a ROM filename.
func (p *Provider) Identify(ctx context.Context, filename string, opts retrometadata.IdentifyOptions) (*retrometadata.GameResult, error) {
	if !p.IsEnabled() {
		return nil, nil
	}

	// Check for MobyGames ID tag in filename
	if match := MobyGamesTagRegex.FindStringSubmatch(filename); len(match) > 1 {
		if id, err := strconv.Atoi(match[1]); err == nil {
			result, err := p.GetByID(ctx, id)
			if err == nil && result != nil {
				return result, nil
			}
		}
	}

	if opts.PlatformID == nil {
		return nil, nil
	}

	var searchTerm string

	// Try Sony serial format for PS1/PS2/PSP platforms
	// MobyGames platform IDs: PS1=6, PS2=7, PSP=46
	platformID := *opts.PlatformID
	if platformID == 6 || platformID == 7 || platformID == 46 {
		if serial := extractSerialCode(filename); serial != "" {
			searchTerm = serial
		}
	}

	// Try Nintendo Switch ID formats (platform ID 203)
	if platformID == 203 && searchTerm == "" {
		_, productID := extractSwitchID(filename)
		if productID != "" {
			searchTerm = productID
		}
	}

	// Try MAME format for arcade platform (ID 143)
	if platformID == 143 && searchTerm == "" {
		if isMAMEFormat(filename) {
			searchTerm = regexp.MustCompile(`\.[^.]+$`).ReplaceAllString(filename, "")
		}
	}

	// Fall back to cleaned filename
	if searchTerm == "" {
		searchTerm = cleanFilename(filename)
	}

	// Search for the game
	params := map[string]string{
		"title":    url.QueryEscape(searchTerm),
		"platform": strconv.Itoa(platformID),
	}

	result, err := p.request(ctx, "/games", params)
	if err != nil {
		return nil, err
	}

	resultMap, ok := result.(map[string]interface{})
	if !ok {
		return nil, nil
	}

	games, ok := resultMap["games"].([]interface{})
	if !ok || len(games) == 0 {
		// Try splitting by special characters
		terms := normalization.SplitSearchTerm(searchTerm)
		if len(terms) > 1 {
			params["title"] = url.QueryEscape(terms[len(terms)-1])
			result, err = p.request(ctx, "/games", params)
			if err != nil {
				return nil, err
			}
			resultMap, ok = result.(map[string]interface{})
			if ok {
				games, _ = resultMap["games"].([]interface{})
			}
		}
	}

	if len(games) == 0 {
		return nil, nil
	}

	// Find best match
	gamesByName := make(map[string]map[string]interface{})
	var names []string
	for _, g := range games {
		if game, ok := g.(map[string]interface{}); ok {
			title := getString(game, "title")
			if title != "" {
				gamesByName[title] = game
				names = append(names, title)
			}
		}
	}

	bestMatch, score := p.FindBestMatch(searchTerm, names)

	if bestMatch != "" {
		if game, ok := gamesByName[bestMatch]; ok {
			gameResult := p.buildGameResult(game)
			gameResult.MatchScore = score
			return gameResult, nil
		}
	}

	return nil, nil
}

// Heartbeat checks if the provider API is accessible.
func (p *Provider) Heartbeat(ctx context.Context) error {
	_, err := p.request(ctx, "/games", map[string]string{"limit": "1"})
	return err
}

func (p *Provider) buildGameResult(game map[string]interface{}) *retrometadata.GameResult {
	providerID := int(getFloat64(game, "game_id"))
	result := &retrometadata.GameResult{
		Provider:    p.Name(),
		ProviderID:  &providerID,
		ProviderIDs: map[string]int{"mobygames": providerID},
		Name:        getString(game, "title"),
		Summary:     getString(game, "description"),
		RawResponse: game,
	}

	// Extract cover URL
	if sampleCover, ok := game["sample_cover"].(map[string]interface{}); ok {
		result.Artwork.CoverURL = getString(sampleCover, "image")
	}

	// Extract screenshots
	if screenshots, ok := game["sample_screenshots"].([]interface{}); ok {
		for _, s := range screenshots {
			if sMap, ok := s.(map[string]interface{}); ok {
				if imgURL := getString(sMap, "image"); imgURL != "" {
					result.Artwork.ScreenshotURLs = append(result.Artwork.ScreenshotURLs, imgURL)
				}
			}
		}
	}

	// Extract metadata
	result.Metadata = p.extractMetadata(game)

	return result
}

func (p *Provider) extractMetadata(game map[string]interface{}) retrometadata.GameMetadata {
	metadata := retrometadata.GameMetadata{
		RawData: game,
	}

	// Genres
	if genres, ok := game["genres"].([]interface{}); ok {
		for _, g := range genres {
			if gMap, ok := g.(map[string]interface{}); ok {
				if name := getString(gMap, "genre_name"); name != "" {
					metadata.Genres = append(metadata.Genres, name)
				}
			}
		}
	}

	// Alternative names
	if altTitles, ok := game["alternate_titles"].([]interface{}); ok {
		for _, t := range altTitles {
			if tMap, ok := t.(map[string]interface{}); ok {
				if title := getString(tMap, "title"); title != "" {
					metadata.AlternativeNames = append(metadata.AlternativeNames, title)
				}
			}
		}
	}

	// Platforms
	if platforms, ok := game["platforms"].([]interface{}); ok {
		for _, pl := range platforms {
			if plMap, ok := pl.(map[string]interface{}); ok {
				platform := retrometadata.Platform{
					Name:        getString(plMap, "platform_name"),
					ProviderIDs: map[string]int{"mobygames": int(getFloat64(plMap, "platform_id"))},
				}
				metadata.Platforms = append(metadata.Platforms, platform)
			}
		}
	}

	// Rating (MobyGames scores are out of 10, convert to 100)
	if mobyScore := getFloat64(game, "moby_score"); mobyScore > 0 {
		rating := mobyScore * 10
		metadata.TotalRating = &rating
	}

	return metadata
}

// GetPlatform returns platform information for a slug.
func (p *Provider) GetPlatform(slug string) *retrometadata.Platform {
	platformSlug := platform.Slug(slug)
	platformID := platform.GetMobyGamesPlatformID(platformSlug)
	if platformID == nil {
		return nil
	}

	name := MobyGamesPlatformNames[*platformID]
	if name == "" {
		name = strings.ReplaceAll(slug, "-", " ")
	}

	return &retrometadata.Platform{
		Slug:        slug,
		Name:        name,
		ProviderIDs: map[string]int{"mobygames": *platformID},
	}
}

func extractSerialCode(filename string) string {
	// Try PS2 OPL format first (SLUS_123.45)
	if match := PS2OPLRegex.FindStringSubmatch(filename); len(match) > 3 {
		prefix := strings.ToUpper(match[1])
		part1 := match[2]
		part2 := match[3]
		return fmt.Sprintf("%s-%s%s", prefix, part1, part2)
	}

	// Try standard Sony serial format (SLUS-12345 or SLUS_12345)
	if match := SonySerialRegex.FindStringSubmatch(filename); len(match) > 2 {
		prefix := strings.ToUpper(match[1])
		number := match[2]
		return fmt.Sprintf("%s-%s", prefix, number)
	}

	return ""
}

func extractSwitchID(filename string) (titleID, productID string) {
	// Try titleID format (16-char hex)
	if match := SwitchTitleDBRegex.FindStringSubmatch(filename); len(match) > 1 {
		titleID = strings.ToUpper(match[1])
	}

	// Try productID format (LA-H-AAAAA)
	if match := SwitchProductIDRegex.FindStringSubmatch(filename); len(match) > 1 {
		productID = strings.ToUpper(match[1])
	}

	return titleID, productID
}

func isMAMEFormat(filename string) bool {
	// Remove extension first
	name := regexp.MustCompile(`\.[^.]+$`).ReplaceAllString(filename, "")
	// MAME names are typically short (under 20 chars) and alphanumeric
	if len(name) > 20 {
		return false
	}
	return MAMEArcadeRegex.MatchString(name)
}

func cleanFilename(filename string) string {
	// Remove extension
	name := regexp.MustCompile(`\.[^.]+$`).ReplaceAllString(filename, "")
	// Remove common tags like (USA), [!], etc.
	name = regexp.MustCompile(`\s*[\(\[][^\)\]]*[\)\]]`).ReplaceAllString(name, "")
	return strings.TrimSpace(name)
}

func getString(m map[string]interface{}, key string) string {
	if v, ok := m[key]; ok {
		if s, ok := v.(string); ok {
			return s
		}
	}
	return ""
}

func getFloat64(m map[string]interface{}, key string) float64 {
	if v, ok := m[key]; ok {
		if f, ok := v.(float64); ok {
			return f
		}
	}
	return 0
}

// MobyGamesPlatformNames maps MobyGames platform IDs to names
var MobyGamesPlatformNames = map[int]string{
	1:   "Linux",
	2:   "DOS",
	3:   "Windows",
	4:   "PC Booter",
	5:   "Windows 3.x",
	6:   "PlayStation",
	7:   "PlayStation 2",
	8:   "Dreamcast",
	9:   "Nintendo 64",
	10:  "Game Boy",
	11:  "Game Boy Color",
	12:  "Game Boy Advance",
	13:  "Xbox",
	14:  "GameCube",
	15:  "SNES",
	16:  "Genesis",
	17:  "Jaguar",
	18:  "Lynx",
	19:  "Amiga",
	20:  "SEGA CD",
	21:  "SEGA 32X",
	22:  "NES",
	23:  "Saturn",
	24:  "Atari ST",
	25:  "Game Gear",
	26:  "SEGA Master System",
	27:  "Commodore 64",
	28:  "Atari 2600",
	29:  "ColecoVision",
	30:  "Intellivision",
	31:  "Apple II",
	32:  "N-Gage",
	33:  "Atari 5200",
	34:  "Atari 7800",
	35:  "3DO",
	36:  "Neo Geo",
	37:  "Vectrex",
	38:  "Virtual Boy",
	39:  "Atari 8-bit",
	40:  "TurboGrafx-16",
	41:  "ZX Spectrum",
	44:  "Nintendo DS",
	46:  "PSP",
	48:  "WonderSwan",
	49:  "WonderSwan Color",
	69:  "Xbox 360",
	74:  "Macintosh",
	81:  "PlayStation 3",
	82:  "Wii",
	86:  "iOS",
	91:  "Android",
	101: "Nintendo 3DS",
	105: "PlayStation Vita",
	132: "Wii U",
	141: "PlayStation 4",
	142: "Xbox One",
	143: "Arcade",
	203: "Nintendo Switch",
	288: "PlayStation 5",
	289: "Xbox Series X|S",
}

func init() {
	// Register the provider factory
	retrometadata.RegisterProvider("mobygames", func(config retrometadata.ProviderConfig, c cache.Cache) (retrometadata.Provider, error) {
		return NewProvider(config, c)
	})
}
